external-crates/move/crates/move-analyzer/editors/code/src/context.ts

```diff
@@ -177,25 +203,176 @@ export class Context {
+     *   - if standalone move-analyzer binary exists, it is now the highest
+     *     version available
+     *   - if CLI binary is available and its version is higher than the currently
+     *     highest one, it is now the highest version
+     *   - if there is a binary bundled with the extension and its version
+     *     is higher than the currently highest one, it is now the highest version,
+     *     and the bundled binary should be installed
+     * - if there is no binary available at this point, report an error
+     *   to the user
+     *
+     * @returns `true` if server binary installation succeeded, `false` otherwise.
+     */
+    async installServerBinary(extensionContext: vscode.ExtensionContext): Promise<boolean> {
+        log.info('Installing language server');
+        // Args to run move-analyzer by invoking server binary
+        const serverArgs: string[] = [];
+        // Args to run move-analyzer by invoking CLI binary
+        const cliArgs = ['analyzer'];
+
+        // Args to get version out of the server binary
+        const serverVersionArgs = serverArgs.concat(['--version']);
+        // Args to get version out of CLI binary
+        const cliVersionArgs = cliArgs.concat(['--version']);
+
+        // Check if server binary is bundled with the extension
+        const bundledServerPath = vscode.Uri.joinPath(extensionContext.extensionUri,
+                                                    'language-server',
+                                                    this.configuration.serverName);
+        const bundledVersion = semanticVersion(bundledServerPath.fsPath, serverVersionArgs);
+        log.info(`bundled version: ${bundledVersion}`);
+
+        const standaloneVersion = semanticVersion(this.configuration.serverPath, serverVersionArgs);
+        log.info(`standalone version: ${standaloneVersion}`);
+
+        const cliVersion = semanticVersion(this.configuration.suiPath, cliVersionArgs);
+        log.info(`cli version: ${cliVersion}`);
+
+        if (this.configuration.serverPath !== this.configuration.defaultServerPath.fsPath) {
+            // User has overwritten default server path (need to compare paths as comparing URIs fails
+            // for some reason even if one is initialized from another).
+            if (standaloneVersion === null) {
+                // The server binary on user-overwritten path does not return version number.
+                // Need to use modal messages, otherwise the promise returned from is not resolved
+                // (and the extension blocks), which can confusing.
+                const items: vscode.MessageItem = { title: 'OK', isCloseAffordance: true };
+                await vscode.window.showInformationMessage(
+                    `The move-analyzer binary at the user-specified path ('${this.configuration.serverPath}') ` +
+                    'is not working. Put the corrent binary in the path or reset user settings and ' +
```

Comment:
```suggestion
                    'is not working. Put the correct binary in the path or reset user settings and ' +
```

Reply:
Which is the right binary and which is the right path? Can the error message provide some direction there?

Reply:
> Which is the right binary and which is the right path? Can the error message provide some direction there?

Good point! I followed a similar path and directed the user to consult extension's README file for troubleshooting